### DISCUSSION TITLE
Shapelet training loss not decreasing

### DISCUSSION THREAD

**Discussion by zillur01** [tslearn-team/tslearn#626] [Q&A]
I have a dataset of 100 time-series. Each is of ~1000x3 dimension. I plotted those time-series and found 4 subsequence patterns. Usually they are of 300x3 dimension. So I manually created the subsequence pattern class and created my training set. 

Now during training loss is not decreasing and accuracy not increasing. also before training starts, it waits for a long time and once training starts, it completes very quickly.

```
Epoch 1/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 532ms/step - categorical_accuracy: 0.2353 - categorical_crossentropy: 1.5912 - loss: 1.5977
Epoch 2/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 27ms/step - categorical_accuracy: 0.2353 - categorical_crossentropy: 1.5864 - loss: 1.5929
Epoch 3/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 28ms/step - categorical_accuracy: 0.2353 - categorical_crossentropy: 1.5817 - loss: 1.5882
Epoch 4/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step - categorical_accuracy: 0.2353 - categorical_crossentropy: 1.5771 - loss: 1.5835
Epoch 5/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step - categorical_accuracy: 0.2353 - categorical_crossentropy: 1.5724 - loss: 1.5789
Epoch 6/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step - categorical_accuracy: 0.2353 - categorical_crossentropy: 1.5678 - loss: 1.5743
Epoch 7/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step - categorical_accuracy: 0.2353 - categorical_crossentropy: 1.5632 - loss: 1.5697
Epoch 8/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step - categorical_accuracy: 0.2353 - categorical_crossentropy: 1.5587 - loss: 1.5651
Epoch 9/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 24ms/step - categorical_accuracy: 0.2647 - categorical_crossentropy: 1.5541 - loss: 1.5606
Epoch 10/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 24ms/step - categorical_accuracy: 0.2647 - categorical_crossentropy: 1.5496 - loss: 1.5561
```

this is my training code:

```
def model(X_train, y_train):
    # Normalize each of the timeseries in the Trace dataset
    X_train = TimeSeriesScalerMinMax().fit_transform(X_train)
    
    # 4 shapelets, each of size 300 timsteps
    shapelet_sizes = {300:4}

    # Initialize the LearningShapelets model
    shp_clf = LearningShapelets(n_shapelets_per_size=shapelet_sizes,
                                optimizer='adam',
                                weight_regularizer=0.001,
                                max_iter=10,
                                verbose=1,
                                random_state=999)

    # Fit the model
    shp_clf.fit(X_train, y_train)
    shp_clf.to_json('shapelet_model.json')
    #shp_clf = LearningShapelets.from_json('shapelet_model.json')

    return shp_clf
```

I am not sure what I am missing here. Could it be the shapelet dimension? I thought since there re 4 classes, and each class is of around 300 timestamps, it should be {300:4}.

**Comment by charavelg** [comment]
According to your logs, loss is actually decreasing( from 1.5977 to 1.5561). The number of training iterations is limited to 10 (max_iter parameter of the LearningShapelets constructor), you might want to increase it, which will mechanically increase the training duration.

Regarding the time spent before training starts, it seems that the shapelets initialization is slow. I'll have a look.

Hope this helps!

**Comment by charavelg** [comment]
Shapelets initialization has been improved on dev branch, you might want to have a look. 
Use `pip install --force-reinstall 'tslearn[all_features] @ https://github.com/tslearn-team/tslearn/archive/main.zip'` to update your env with lastest version
